Changed __str__ return to include Tensor-Wise information for issue #1439.

neel2299 opened the pull request:
## 🚀 🚀 Pull Request

### Checklist:

- [x]  [My code follows the style guidelines of this project](https://www.notion.so/activeloop/Engineering-Guidelines-d6e502306d0e4133a8ca507516d1baab) and the [Contributing document](https://github.com/activeloopai/Hub/blob/release/2.0/CONTRIBUTING.md)
- [x]  I have commented my code, particularly in hard-to-understand areas
- [ ]  I have kept the `coverage-rate` up
- [x]  I have performed a self-review of my own code and resolved any problems
- [x]  I have checked to ensure there aren't any other open [Pull Requests](https://github.com/activeloopai/Hub/pulls) for the same change
- [ ]  I have described and made corresponding changes to the relevant documentation
- [x]  New and existing unit tests pass locally with my changes


### Changes

Used pandas to_string method to get the tensor information into a table.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/activeloopai/Hub?pullRequest=1543) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/activeloopai/Hub?pullRequest=1543) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/activeloopai/Hub?pullRequest=1543) it.</sub>

farizrahman4u commented on hub/core/dataset/dataset.py:
```diff
@@ -1,5 +1,5 @@
 # type: ignore
-from unittest import skip
+from pandas import DataFrame
```
Lets not have a Pandas dependency just for __str__

neel2299 replied:
Okay, Let me try implementing the to_string method then. Hope it works.

farizrahman4u reviewed (changes requested):
This PR reverts hidden tensors and can't be accepted in the current state.

neel2299 commented:
> This PR reverts hidden tensors and can't be accepted in the current state.


For some reason, I noticed this comment only now, Can you please point me to the documentation for hidden tensors? I didn't know that my commit was affecting something.

farizrahman4u commented on hub/core/dataset/dataset.py:
```diff
@@ -355,24 +379,18 @@ def create_tensor(
         meta: DatasetMeta = self.meta
         ffw_dataset_meta(meta)
-        meta.add_tensor(name, hidden=hidden)
+        meta.add_tensor(name)
```
Dont change this.

farizrahman4u commented on hub/core/dataset/dataset.py:
```diff
@@ -470,7 +488,7 @@ def delete_group(self, name: str, large_ok: bool = False):
             tensors = [
                 posixpath.join(name, tensor)
-                for tensor in self[name]._all_tensors_filtered(include_hidden=True)
+                for tensor in self[name]._all_tensors_filtered
```
Dont change this.

farizrahman4u commented on hub/core/dataset/dataset.py:
```diff
@@ -1091,30 +1139,25 @@ def _ungrouped_tensors(self) -> Dict[str, Tensor]:
 
-    def _all_tensors_filtered(self, include_hidden: bool = True) -> List[str]:
+    @property
+    def _all_tensors_filtered(self) -> List[str]:
```
Dont change this.

farizrahman4u commented on hub/core/dataset/dataset.py:
```diff
@@ -1091,30 +1139,25 @@ def _ungrouped_tensors(self) -> Dict[str, Tensor]:
             for t in self.version_state["full_tensors"]
-            if (not self.group_index or t.startswith(self.group_index + "/"))
-            and (include_hidden or t not in hidden_tensors)
+            if not self.group_index or t.startswith(self.group_index + "/")
```
Dont change this.

farizrahman4u commented on hub/core/dataset/dataset.py:
```diff
@@ -1091,30 +1139,25 @@ def _ungrouped_tensors(self) -> Dict[str, Tensor]:
 
-    def _tensors(self, include_hidden: bool = True) -> Dict[str, Tensor]:
+    @property
+    def tensors(self) -> Dict[str, Tensor]:
```
Dont change this.

farizrahman4u commented on hub/core/dataset/dataset.py:
```diff
@@ -1091,30 +1139,25 @@ def _ungrouped_tensors(self) -> Dict[str, Tensor]:
                 self.index
             ]
-            for t in self._all_tensors_filtered(include_hidden)
+            for t in self._all_tensors_filtered
```
Dont change this.

farizrahman4u commented on hub/core/dataset/dataset.py:
```diff
@@ -1287,7 +1330,7 @@ def rechunk(
 
         if tensors is None:
-            tensors = list(self._tensors())
+            tensors = list(self.tensors.keys())
```
Dont change this.

farizrahman4u commented on hub/core/dataset/dataset.py:
```diff
@@ -1368,16 +1385,14 @@ def append(self, sample: Dict[str, Any], skip_ok: bool = False):
                     )
         for k in sample:
-            if k not in self._tensors():
+            if k not in self.tensors:
```
Dont change this.

farizrahman4u commented on hub/core/dataset/dataset.py:
```diff
@@ -1368,16 +1385,14 @@ def append(self, sample: Dict[str, Any], skip_ok: bool = False):
             Exception: Error while attempting to rollback appends.
         """
-        if isinstance(sample, Dataset):
-            sample = sample.tensors
```
Dont remove this.

neel2299 replied:
Oops, This seems like quite the mess, Let me pull changes from hub main and redo the changes.

neel2299 replied:
Thanks for the review 😄

farizrahman4u reviewed (changes requested):
I have commented on the lines which reverts previous PRs. Seems your git history is messed up.

neel2299 commented:
Hello!!, I have done it afresh, and checked the comparison as well. This time its showing only the lines I have added and removed.(Previously I think I made a mess while trying to manually compare the files with the pip version.)

farizrahman4u reviewed (changes requested):
* Move the helper methods (including the logic in `__str__` to a util file, like `hub/util/summary.py`
* Use snake case everywhere (except class names)
* Instead of changing `__str__`, add `.summary()` method to both `Tensor` and `Dataset`.

neel2299 commented:
@farizrahman4u  Thank you for the review!! I have made the changes requested. Please tell me if it needs more changes.

FayazRahman commented on hub/core/dataset/dataset.py:
```diff
@@ -1138,7 +1142,49 @@ def delete(self, large_ok=False):
+            tensor_name = tensor
+            tensor_htype = tensor_object.htype
+            tensor_shape = str(tensor_object.shape)
+            tensor_dtype = tensor_object.dtype.name
```
`tensor_object.dtype` could be None and it will not have `name` attribute. Just do `ds.create_tensor("abc")` to reproduce the error.

neel2299 replied:
I made the changes. Can you please review it once again? @FayazRahman @farizrahman4u @AbhinavTuli @mikayelh

neel2299 commented:
@FayazRahman Thank You for the review!! I have added support for None types for all attributes except for tensor_shape (It initializes at (0,)). 

Also while trying to add new tensors I noticed that with `ds.create_tensor("abc")` , along with "abc" we get another tensor "_abc_id" . This was happening with the current main hub repo as well. Is it getting created for some specific reason?
![image](https://user-images.githubusercontent.com/61134339/160423154-9f739bbb-e08e-4391-afea-5a398d9be0c3.png)

farizrahman4u reviewed (changes requested):
* Use `self.tensors` instead of `self.version_state["full_tensors"]` to hide hidden tensors.
* Many variables are still in camel case, please switch to snake case as mentioned in the previous review.
*  A simple unit test would be nice.

farizrahman4u commented on hub/core/dataset/dataset.py:
```diff
@@ -1138,7 +1142,57 @@ def delete(self, large_ok=False):
+        divider = ["-------"] * 5
+        tensor_dict = self.version_state[
+            "full_tensors"
+        ]  # Creating a list of tensors in the dataset
```
Comment not required - this is a dict not a list, and should `self.tensors` instead of `self.version_state["full_tensors"]`

neel2299 commented:
Hello!! Thanks alot for the review. I have made the changes. Last time I actually thought that we had to move only the helper functions out to utils(not the ones which were implemented as methods). Sorry about that. Please check out the changes. @farizrahman4u @FayazRahman

farizrahman4u commented:
> Hello!! Thanks alot for the review. I have made the changes. Last time I actually thought that we had to move only the helper functions out to utils(not the ones which were implemented as methods). Sorry about the late reply, had my end terms going. I will write unit tests for both tensor and dataset **str** methods soon and if its alright in another PR. Please check out the new changes. @farizrahman4u @FayazRahman

Tests need to in the same PR @neel2299 . Just calling `.summary` on tensors and datasets with htype specified and unspecified, compression specified and unspecified etc would do.

neel2299 commented:
Thank you for the review :). I changed the previous test_stringify in test_api.py according to the new implementation. Please tell me if there are any other changes to be made. @farizrahman4u @FayazRahman

neel2299 commented:
Thanks for the review once again :) . I changed the tests for test_stringify according to the new implementation and added summary tests. @farizrahman4u @FayazRahman

neel2299 commented:
Added test for local path (the failed test) @farizrahman4u @FayazRahman

neel2299 commented:
Learned alot!! @tatevikh . Thanks to community :)
